pkg/informer/informer.go

```diff
@@ -58,14 +59,64 @@ const (
 	byGroupVersionResourceIndex = "byGroupVersionResource"
 )
 
-// DynamicDiscoverySharedInformerFactory is a SharedInformerFactory that
+// GVRPartialMetadata provides the required metadata about a GVR
+// for which an informer should be started.
+type GVRPartialMetadata struct {
+	Names apiextensionsv1.CustomResourceDefinitionNames
+	Scope apiextensionsv1.ResourceScope
+}
+
+// GVRSource is a source of information about available GVRs,
+// and provides a way to register for changes in the available GVRs (addition or removal)
+// so that the list of informers can be updated.
+type GVRSource interface {
+	// GVRs returns the required metadata about all GVRs known by the GVRSource
+	GVRs() map[schema.GroupVersionResource]GVRPartialMetadata
+
+	// Synced returns true if informers used by this GVRSource are synced
+	Synced() bool
+
+	// Changes returns a channel to which the GVRSource writes whenever
+	// its list of known GVRs has changed
+	Changes() <-chan struct{}
+
+	// FilterBuiltinGVRs is a function that allows filtering out
+	// builtin GVRs that should not be considered.
+	// The function should return false to ignore a builtin GVR,
+	// and true to keep it.
+	FilterBuiltinGVRs(gvr schema.GroupVersionResource) bool
+}
+
+// genericInformerBase is a base interface that would be satisfied by both
+// cluster-aware and cluster-unaware generic informers
+type genericInformerBase[Informer cache.SharedIndexInformer, Lister genericListerBase] interface {
+	// Informer is the shared index informer returned by the generic informer.
+	// It would be either a SharedIndexInformer or a ScopeableSharedIndexInformer,
+	// according to the cluster-awareness.
+	Informer() Informer
+
+	// Lister is the generic lister returned by the generic informer.
+	// It would be either a GenericLister or a GenericClusterLister,
+	// according to the cluster-awareness
+	Lister() Lister
+}
+
+type genericListerBase interface {
+	List(selector labels.Selector) (ret []runtime.Object, err error)
+}
+
+var _ genericInformerBase[kcpcache.ScopeableSharedIndexInformer, kcpcache.GenericClusterLister] = (kcpinformers.GenericClusterInformer)(nil)
+var _ genericInformerBase[cache.SharedIndexInformer, cache.GenericLister] = (informers.GenericInformer)(nil)
+
+// GenericDiscoveringDynamicSharedInformerFactory is a SharedInformerFactory that
 // dynamically discovers new types and begins informing on them.
-type DynamicDiscoverySharedInformerFactory struct {
-	dynamicClient     kcpdynamic.ClusterInterface
-	filterFunc        func(interface{}) bool
-	indexers          cache.Indexers
-	crdIndexer        cache.Indexer
-	crdInformerSynced cache.InformerSynced
+// It is a generic implementation for both logical-cluster-aware and
+// logical-cluster-unaware (== single cluster, standard kube) informers.
+type GenericDiscoveringDynamicSharedInformerFactory[Informer cache.SharedIndexInformer, Lister genericListerBase, GenericInformer genericInformerBase[Informer, Lister]] struct {
```

Comment:
why does this need to be parametrized by informers and listers? This makes everything here much more complicated.

Reply:
clarified via Slack. Still kind of ugly, but ok.